File: renderer/manim_renderer/project.py
import json
from pathlib import Path
import tempfile

from jsonschema import Draft202012Validator
from .elements.expressions import check_expression
from .timeline import validate_timeline, frame_at


SCHEMA_PATH = Path(__file__).resolve().parents[2] / "contracts/project.schema.json"
SCHEMA = json.loads(SCHEMA_PATH.read_text(encoding="utf-8"))
Draft202012Validator.check_schema(SCHEMA)
VALIDATOR = Draft202012Validator(SCHEMA)


def validate_project(project: dict) -> None:
    json.dumps(project, allow_nan=False)
    for error in VALIDATOR.iter_errors(project):
        path = "/".join(str(part) for part in error.absolute_path)
        raise ValueError(f"{path or '/'}: {error.message}")
    scene = project["scene"]
    if frame_at(scene["durationMs"]) < 1:
        raise ValueError("The scene must span at least one frame (67 ms).")
    for element_id, element in scene["elements"].items():
        if element["appearsAtMs"] >= scene["durationMs"]:
            raise ValueError(f"elements/{element_id}/appearsAtMs: must precede scene end")
        if frame_at(element["appearsAtMs"]) >= frame_at(scene["durationMs"]):
            raise ValueError("The element must appear before the final frame.")
        if element["kind"] in ("line", "arrow") and element["position"] == element["end"]:
            raise ValueError("Lines and arrows require distinct endpoints.")
        end = element.get("disappearsAtMs", scene["durationMs"])
        if end > scene["durationMs"] or frame_at(end) <= frame_at(element["appearsAtMs"]):
            raise ValueError("The element must end after it starts and within the scene.")
        for key in ("xRange", "yRange"):
            if key in element:
                low, high = element[key]
                if low >= high or low < -100 or high > 100 or high - low < 0.1:
                    raise ValueError("Invalid axis range: the minimum must be less than the maximum, between −100 and 100.")
        if element["kind"] == "functionGraph":
            check_expression(element["expression"], element["xRange"])
            if "axesId" in element:
                axes = scene["elements"].get(element["axesId"])
                if not axes or axes["kind"] not in ("axes", "numberPlane"):
                    raise ValueError("Select existing axes for the graph.")
                if axes.get("scale", 1) != 1:
                    raise ValueError("Adjust linked axes width and height, keeping their scale at 1.")
        if element["kind"] == "areaUnderGraph":
            graph = scene["elements"].get(element["graphId"])
            if not graph or graph["kind"] != "functionGraph":
                raise ValueError("Select an existing graph for the area.")
            if "axesId" not in graph:
                raise ValueError("The area requires a graph linked to axes.")
            if element["xRange"][0] < graph["xRange"][0] or element["xRange"][1] > graph["xRange"][1]:
                raise ValueError("The area range must stay within the graph range.")
    validate_timeline(scene)


def load_project(path: Path) -> dict:
    project = json.loads(path.read_text(encoding="utf-8"))
    validate_project(project)
    return project


def save_project(project: dict, path: Path) -> None:
    validate_project(project)
    contents = json.dumps(project, indent=2, ensure_ascii=False, allow_nan=False) + "\n"
    with tempfile.NamedTemporaryFile(
        mode="w", encoding="utf-8", dir=path.parent, delete=False
    ) as temporary:
        temporary_path = Path(temporary.name)
        try:
            temporary.write(contents)
        except BaseException:
            temporary.close()
            temporary_path.unlink(missing_ok=True)
            raise
    try:
        temporary_path.replace(path)
    finally:
        temporary_path.unlink(missing_ok=True)

File: src/domain/project.ts
import validate from "../generated/validate.js";
import type { ManimEditorProject, MathTex } from "../generated/project";
import { validateTimeline } from "./timeline";
import { expressionError } from "./expression";
import example from "../../examples/equation.json";

export type Project = ManimEditorProject;
export type Element = Project["scene"]["elements"][string];
export type { Animation } from "../generated/project";
export type Equation = MathTex;
export const initialProject = (): Project => {
  const candidate: unknown = structuredClone(example);
  const error = validateProject(candidate);
  if (error) throw new Error(error);
  return candidate as Project;
};

export function validateProject(project: unknown): string | null {
  if (!validate(project))
    return (
      "Invalid project: " +
      (validate.errors?.[0]?.message ?? "check the fields.")
    );
  const scene = (project as Project).scene;
  for (const element of Object.values(scene.elements)) {
    for (const value of Object.values(element))
      if ((typeof value === "number" && !Number.isFinite(value)) ||
        (Array.isArray(value) && value.some((item) => !Number.isFinite(item))))
        return "All element numbers must be finite.";
    if (element.appearsAtMs >= scene.durationMs)
      return "The element must appear before the scene ends.";
    const end = element.disappearsAtMs ?? scene.durationMs;
    if (end > scene.durationMs || Math.round(end * 15 / 1000) <= Math.round(element.appearsAtMs * 15 / 1000))
      return "The element must end after it starts and within the scene.";
    if ((element.kind === "line" || element.kind === "arrow") && element.position.every((v, i) => v === element.end[i]))
      return "Lines and arrows require distinct endpoints.";
    for (const range of [("xRange" in element ? element.xRange : undefined), ("yRange" in element ? element.yRange : undefined)])
      if (range && (range[0] >= range[1] || range[0] < -100 || range[1] > 100 || range[1] - range[0] < 0.1))
        return "Invalid axis range: the minimum must be less than the maximum, between −100 and 100.";
    if (element.kind === "functionGraph") {
      const invalid = expressionError(element.expression, element.xRange);
      if (invalid) return invalid;
      if (element.axesId) {
        const axes = scene.elements[element.axesId];
        if (!axes || !["axes", "numberPlane"].includes(axes.kind)) return "Select existing axes for the graph.";
        if ((axes.scale ?? 1) !== 1) return "Adjust linked axes width and height, keeping their scale at 1.";
      }
    }
    if (element.kind === "areaUnderGraph") {
      const graph = scene.elements[element.graphId];
      if (!graph || graph.kind !== "functionGraph") return "Select an existing graph for the area.";
      if (!graph.axesId) return "The area requires a graph linked to axes.";
      if (element.xRange[0] < graph.xRange[0] || element.xRange[1] > graph.xRange[1])
        return "The area range must stay within the graph range.";
    }
  }
  return validateTimeline(scene);
}
